rs/state_layout/src/state_layout.rs

```diff
@@ -1604,29 +1664,108 @@ impl<Permissions: AccessPolicy> CanisterLayout<Permissions> {
         PageMapLayout {
-            canister_root: self.canister_root.clone(),
+            root: self.canister_root.clone(),
+            name_stem: "vmemory_0".into(),
+            permissions_tag: PhantomData,
+        }
+    }
+
+    pub fn stable_memory(&self) -> PageMapLayout<Permissions> {
+        PageMapLayout {
+            root: self.canister_root.clone(),
+            name_stem: "stable_memory".into(),
+            permissions_tag: PhantomData,
+        }
+    }
+
+    pub fn wasm_chunk_store(&self) -> PageMapLayout<Permissions> {
+        PageMapLayout {
+            root: self.canister_root.clone(),
+            name_stem: "wasm_chunk_store".into(),
+            permissions_tag: PhantomData,
+        }
+    }
+}
+
+pub struct SnapshotLayout<Permissions: AccessPolicy> {
+    snapshot_root: PathBuf,
+    permissions_tag: PhantomData<Permissions>,
+}
+
+impl<Permissions: AccessPolicy> SnapshotLayout<Permissions> {
+    pub fn new(snapshot_root: PathBuf) -> Result<Self, LayoutError> {
+        Permissions::check_dir(&snapshot_root)?;
+        Ok(Self {
+            snapshot_root,
+            permissions_tag: PhantomData,
+        })
+    }
+
+    pub fn raw_path(&self) -> PathBuf {
+        self.snapshot_root.clone()
+    }
+
+    pub fn wasm(&self) -> WasmFile<Permissions> {
+        self.snapshot_root.join(WASM_FILE).into()
+    }
+
+    pub fn snapshot(
+        &self,
+    ) -> ProtoFileWith<pb_canister_snapshot_bits::CanisterSnapshotBits, Permissions> {
+        self.snapshot_root.join(SNAPSHOT_FILE).into()
+    }
+
+    pub fn vmemory_0(&self) -> PageMapLayout<Permissions> {
+        PageMapLayout {
+            root: self.snapshot_root.clone(),
             name_stem: "vmemory_0".into(),
             permissions_tag: PhantomData,
         }
     }
 
     pub fn stable_memory(&self) -> PageMapLayout<Permissions> {
         PageMapLayout {
-            canister_root: self.canister_root.clone(),
+            root: self.snapshot_root.clone(),
             name_stem: "stable_memory".into(),
             permissions_tag: PhantomData,
         }
     }
 
     pub fn wasm_chunk_store(&self) -> PageMapLayout<Permissions> {
         PageMapLayout {
-            canister_root: self.canister_root.clone(),
+            root: self.snapshot_root.clone(),
             name_stem: "wasm_chunk_store".into(),
             permissions_tag: PhantomData,
         }
     }
 }
 
+impl<P> SnapshotLayout<P>
+where
+    P: WritePolicy,
+{
+    /// Remove the entire directory for the snapshot.
+    pub fn delete_dir(&self) -> Result<(), LayoutError> {
+        let map_error = |err| LayoutError::IoError {
+            path: self.raw_path(),
+            message: "Cannot remove snapshot.".to_string(),
+            io_err: err,
+        };
+
+        std::fs::remove_dir_all(self.raw_path()).map_err(map_error)?;
+
+        // Remove the parent directory named after the canister if this was the last snapshot of that canister.
```

Comment:
The [comment](https://github.com/dfinity/ic/pull/702/files#diff-4d12f2952e788711a5eff1bca6d5c80bbd4848d76b853cdcdb516301389c577bR477) in PR #702 says `snapshot_operations` are not parallelized, which makes sense. Otherwise, trying removing the same parent folder may returns an error.

I am curious if we really care about clean up an empty snapshots folder. For example, in `filter_tip_canisters`, even if we find all canisters are removed, we still keep the empty canisters folder.

Reply:
I just saw that this behavior is taken into account in the tests where we count the number of canisters under `SNAPSHOTS_DIR`.